src/main/c/Region.cpp

```diff
@@ -0,0 +1,111 @@
+#include "hail/Region.h"
+#include "hail/NativeObj.h"
+#include "hail/NativePtr.h"
+#include <jni.h>
+
+namespace hail {
+
+void Region::clear_but_keep_mem() {
+  pos_ = (chunk_.empty() ? kChunkCap : 0);
+  for (auto& full : full_chunks_) free_chunks_.emplace_back(std::move(full));
+  full_chunks_.clear();
+  // Move big_used_ to big_free_
+  for (auto& used : big_used_) big_free_.emplace_back(std::move(used));
+  big_used_.clear();
+  // Sort in descending order of size
+  std::sort(
+    big_free_.begin(),
+    big_free_.end(),
+    [](const UniqueChunk& a, const UniqueChunk& b)->bool { return (a.size_ > b.size_); }
+  );
+  if (big_free_.size() > kNumBigToKeep) big_free_.resize(kNumBigToKeep);
+}
+
+void Region::new_chunk() {
+  if (!chunk_.empty()) full_chunks_.emplace_back(std::move(chunk_));
```

Comment:
`full_chunks_.emplace_back(std::move(chunk_));` works, but just so you know `full_chunks_.push_back(std::move(chunk_));` will do the exact same thing (and there's no chance of accidentally calling any constructor other than the move constructor).

Reply:
I'm always going to use emplace_back when I can to make it explicit that there's no temporary object.